In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
from mlxtend.frequent_patterns import apriori, fpgrowth, fpmax, hmine
from mlxtend.frequent_patterns import association_rules

import sys
sys.path.append('C:/Users/Owner/Downloads/dhs_service_records_synthesized_final.csv/')  # Adjust the path to where dhs_util.py is located

import dhs_util
from dhs_util import *

# Change to the directory where the file is located
os.chdir('C:/Users/Owner/Downloads/dhs_service_records_synthesized_final.csv/')

# Load the dataset and apply the necessary functions
df = pd.read_csv('dhs_service_records_synthesized_final.csv')
df = dhs_preprocessing(df)
df, service_map = add_service_label(df)
df = add_age_bin(df)

# Get recipient attributes
recipient = get_recipient_attribute(df)




In [2]:
df_pivot = df.pivot_table(index='id', columns='service', aggfunc='size', fill_value=0, observed=False)
df_pivot.head()


service  Children_Attending_Early_Childhood_Programs_Managed_by_DHS  \
id                                                                    
1                                                        0            
2                                                        0            
3                                                        0            
4                                                        0            
5                                                        0            

service  Children_Receiving_Child_Welfare_Services  \
id                                                   
1                                                0   
2                                                0   
3                                                0   
4                                                0   
5                                                0   

service  Children_Receiving_DHS_Funded_Out_of_School_Programs  \
id                                                              
1                                                        0      
2                                                        0      
3                                                        0      
4                                                        0      
5                                                        0      

service  Children_Receiving_Early_Intervention_Services  Children_in_Care  \
id                                                                          
1                                                     0                 0   
2                                                     0                 0   
3                                                     0                 0   
4                                                     0                 0   
5                                                     0                 0   

service  Families_Receiving_Child_Welfare_Services  Homeless_Population  \
id                                                                        
1                                                0                    0   
2                                                0                    0   
3                                                0                    0   
4                                                0                    0   
5                                                0                    0   

service  Homicides  Individuals_Receiving_DHS_Services  \
id                                                       
1                0                                   0   
2                0                                   0   
3                0                                   0   
4                0                                   0   
5                0                                   0   

service  Individuals_Receiving_Family_Strengthening_Programs  ...  \
id                                                            ...   
1                                                        0    ...   
2                                                        0    ...   
3                                                        0    ...   
4                                                        0    ...   
5                                                        0    ...   

service  Individuals_Receiving_Intellectual_Disability_Services  \
id                                                                
1                                                        0        
2                                                        0        
3                                                        0        
4                                                        0        
5                                                        0        

service  Individuals_Receiving_Mental_Health_Services  \
id                                                      
1                                                   0   
2                                                   0   
3           

## Homework 3

In [20]:
# Convert to one-hot encoding for services
df_pivot_binary = df_pivot.applymap(lambda x: 1 if x > 0 else 0)

# Check if the conversion worked
df_pivot_binary.head()


C:\Users\Owner\AppData\Local\Temp\ipykernel_12628\2692346075.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_pivot_binary = df_pivot.applymap(lambda x: 1 if x > 0 else 0)


service  Children_Attending_Early_Childhood_Programs_Managed_by_DHS  \
id                                                                    
1                                                        0            
2                                                        0            
3                                                        0            
4                                                        0            
5                                                        0            

service  Children_Receiving_Child_Welfare_Services  \
id                                                   
1                                                0   
2                                                0   
3                                                0   
4                                                0   
5                                                0   

service  Children_Receiving_DHS_Funded_Out_of_School_Programs  \
id                                                              
1                                                        0      
2                                                        0      
3                                                        0      
4                                                        0      
5                                                        0      

service  Children_Receiving_Early_Intervention_Services  Children_in_Care  \
id                                                                          
1                                                     0                 0   
2                                                     0                 0   
3                                                     0                 0   
4                                                     0                 0   
5                                                     0                 0   

service  Families_Receiving_Child_Welfare_Services  Homeless_Population  \
id                                                                        
1                                                0                    0   
2                                                0                    0   
3                                                0                    0   
4                                                0                    0   
5                                                0                    0   

service  Homicides  Individuals_Receiving_DHS_Services  \
id                                                       
1                0                                   0   
2                0                                   0   
3                0                                   0   
4                0                                   0   
5                0                                   0   

service  Individuals_Receiving_Family_Strengthening_Programs  ...  \
id                                                            ...   
1                                                        0    ...   
2                                                        0    ...   
3                                                        0    ...   
4                                                        0    ...   
5                                                        0    ...   

service  Individuals_Receiving_Intellectual_Disability_Services  \
id                                                                
1                                                        0        
2                                                        0        
3                                                        0        
4                                                        0        
5                                                        0        

service  Individuals_Receiving_Mental_Health_Services  \
id                                                      
1                                                   0   
2                                                   0   
3           

In [5]:
from mlxtend.frequent_patterns import apriori

# Use Apriori with support > 0.01
frequent_itemsets_apriori = apriori(df_pivot_binary, min_support=0.01, use_colnames=True)
frequent_itemsets_apriori.sort_values(by='support', ascending=False).head()


C:\Users\Owner\anaconda3\Lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                                           itemsets
4   0.941422            (Individuals_Receiving_Income_Supports)
2   0.153844               (Individuals_Receiving_DHS_Services)
14  0.139131  (Individuals_Receiving_DHS_Services, Individua...
15  0.103528  (Individuals_Receiving_Mental_Health_Services,...
6   0.103528     (Individuals_Receiving_Mental_Health_Services)

In [6]:
from mlxtend.frequent_patterns import fpgrowth

# Use fpgrowth with support > 0.0001
frequent_itemsets_fpgrowth = fpgrowth(df_pivot_binary, min_support=0.0001, use_colnames=True)
frequent_itemsets_fpgrowth.sort_values(by='support', ascending=False).head()


C:\Users\Owner\anaconda3\Lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                                           itemsets
0   0.941422            (Individuals_Receiving_Income_Supports)
1   0.153844               (Individuals_Receiving_DHS_Services)
22  0.139131  (Individuals_Receiving_DHS_Services, Individua...
3   0.103528     (Individuals_Receiving_Mental_Health_Services)
30  0.103528  (Individuals_Receiving_Mental_Health_Services,...

In [7]:
from mlxtend.frequent_patterns import fpmax

# Use fpmax with support > 0.0001 
frequent_itemsets_fpmax = fpmax(df_pivot_binary, min_support=0.0001, use_colnames=True)
frequent_itemsets_fpmax.sort_values(by='support', ascending=False).head()


C:\Users\Owner\anaconda3\Lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                                           itemsets
56  0.000326  (Individuals_Receiving_Substance_Use_Disorder_...
1   0.000305  (Individuals_Receiving_Income_Supports, Childr...
34  0.000253  (Individuals_Receiving_DHS_Services, Mental_He...
63  0.000247  (Individuals_Receiving_DHS_Services, Individua...
11  0.000223  (Individuals_Receiving_DHS_Services, Children_...

In [8]:
# Fpmax association rules with support > 0.0003 
rules_fpmax = association_rules(frequent_itemsets_fpmax, metric="support", min_threshold=0.0003, support_only=True)
rules_fpmax.head()


antecedents  \
0            (Individuals_Receiving_Income_Supports)   
1  (Children_Attending_Early_Childhood_Programs_M...   
2  (Individuals_Receiving_Substance_Use_Disorder_...   
3  (Individuals_Receiving_Substance_Use_Disorder_...   
4  (Individuals_Receiving_Substance_Use_Disorder_...   

                                         consequents  antecedent support  \
0  (Children_Attending_Early_Childhood_Programs_M...                 NaN   
1            (Individuals_Receiving_Income_Supports)                 NaN   
2                                  (Jail_Population)                 NaN   
3            (Individuals_Receiving_Income_Supports)                 NaN   
4     (Individuals_Receiving_Mental_Health_Services)                 NaN   

   consequent support   support  confidence  lift  leverage  conviction  \
0                 NaN  0.000305         NaN   NaN       NaN         NaN   
1                 NaN  0.000305         NaN   NaN       NaN         NaN   
2                 NaN  0.000326         NaN   NaN       NaN         NaN   
3                 NaN  0.000326         NaN   NaN       NaN         NaN   
4                 NaN  0.000326         NaN   NaN       NaN         NaN   

   zhangs_metric  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN

#### Given the data analysis results, I would suggest that DHS provide an increase in services for Mental Health, Children Attending Early Learning Programs, and Homeless Population in association with Child Welfare Services. 

## In-Class Questions

In [11]:
import pandas as pd

# 1. Extract the first 6 months' data
df['date'] = pd.to_datetime(df['date'])  # Convert to datetime
first_six_months = df[(df['date'].dt.month >= 1) & (df['date'].dt.month <= 6)]

# 2. Select recipients who used Income Support (S15)
income_support_recipients = first_six_months[first_six_months['serv'] == 'S15']['id'].unique()

# 3. Create Group 1 (Only used Income Service) and Group 2 (Used Income Service with other services)
group1_ids = first_six_months.groupby('id').filter(lambda x: (x['serv'] == 'S15').all())['id'].unique()
group2_ids = list(set(income_support_recipients) - set(group1_ids))

# 4. Extract data for the last 3 months (October to December) for Group 1 and Group 2
last_three_months = df[(df['date'].dt.month >= 10) & (df['date'].dt.month <= 12)]
group1_last_three = last_three_months[last_three_months['id'].isin(group1_ids)]
group2_last_three = last_three_months[last_three_months['id'].isin(group2_ids)]

# 5. Analyze drop-off in Income Support (S15) in the last 3 months for both groups
group1_income_dropped = group1_last_three[group1_last_three['serv'] != 'S15']['id'].unique()
group2_income_dropped = group2_last_three[group2_last_three['serv'] != 'S15']['id'].unique()

# Comparison of drop-off
print(f"Group 1 drop-off count (Income Support only): {len(group1_income_dropped)}")
print(f"Group 2 drop-off count (Income Support with other services): {len(group2_income_dropped)}")

# 6. Demographic analysis: Checking if age, gender, or other demographics affect behavior change
group1_last_three_demo = last_three_months[last_three_months['id'].isin(group1_ids)][['id', 'age', 'gender']]
group2_last_three_demo = last_three_months[last_three_months['id'].isin(group2_ids)][['id', 'age', 'gender']]

# Checking age and gender distribution
group1_age_gender = group1_last_three_demo.drop_duplicates(subset=['id']).groupby(['age', 'gender']).size().reset_index(name='count')
group2_age_gender = group2_last_three_demo.drop_duplicates(subset=['id']).groupby(['age', 'gender']).size().reset_index(name='count')

# Print demographic comparison
print("Group 1 Demographics (age, gender):")
print(group1_age_gender)
print("\nGroup 2 Demographics (age, gender):")
print(group2_age_gender)


Group 1 drop-off count (Income Support only): 102
Group 2 drop-off count (Income Support with other services): 10923
Group 1 Demographics (age, gender):
     age    gender  count
0     14    1~Male      0
1     14  2~Female      2
2     14  99~Other      0
3     16    1~Male      2
4     16  2~Female      1
..   ...       ...    ...
151   90  2~Female      0
152   90  99~Other      0
153   93    1~Male      0
154   93  2~Female      1
155   93  99~Other      0

[156 rows x 3 columns]

Group 2 Demographics (age, gender):
     age    gender  count
0     14    1~Male     39
1     14  2~Female     50
2     14  99~Other      0
3     15    1~Male     26
4     15  2~Female     35
..   ...       ...    ...
265  104  2~Female      1
266  104  99~Other      0
267  106    1~Male      2
268  106  2~Female      0
269  106  99~Other      0

[270 rows x 3 columns]


C:\Users\Owner\AppData\Local\Temp\ipykernel_12628\1217822422.py:33: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  group1_age_gender = group1_last_three_demo.drop_duplicates(subset=['id']).groupby(['age', 'gender']).size().reset_index(name='count')
C:\Users\Owner\AppData\Local\Temp\ipykernel_12628\1217822422.py:34: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  group2_age_gender = group2_last_three_demo.drop_duplicates(subset=['id']).groupby(['age', 'gender']).size().reset_index(name='count')


#### Interpretation:

Drop-Off Counts:
In regard to Group One, 102 recipients dropped out of using Income Support Services when they were receiving no other services. Comparatively, Group Two saw a drop of 10, 923 recipients from using Income Support Services. This suggests that recipients using multiple services in conjunction with Income Support were possibly in periods of transition away from needing Income Support or moving away from Income Support by meeting their needs with other services.

Demographic Analysis:
There was a small drop in Group One from the Young and Elderly demographics, but no major peaks in specific age groups. Group Two has a ihgher concentration of younger recipients than the thin, even spread of Group One. The analysis of Group Two showed significantly more drop-offs in terms of young recipients. This set is also larger and more diverse in terms of age and gender.

Gender: 
There is not a distinguishable pattern in gender for drop-offs for group One, while Group 2 is much larger and shows a high drop-off rate for younger female recipients as opposed to male recipients. Which means that the use of other services, changing needs, etc. could be attributed to higher drop-off counts.

I would be curious as to the younger recipients and want to know if they are transitioning to services that are more specialized to their needs.

In [23]:
import time
from mlxtend.frequent_patterns import apriori

# min_freq values to test
min_freq_values = [0.01, 0.005, 0.001, 0.0005, 0.0001]

# Store results
results = []

# Test each min_freq and record execution time and number of itemsets
for min_freq in min_freq_values:
    print(f"Testing min_freq = {min_freq}")
    
    start_time = time.time()  # Start timing
    
    # Run apriori with current min_freq
    frequent_itemsets = apriori(df_pivot_binary, min_support=min_freq, use_colnames=True)
    
    end_time = time.time()  # End timing
    
    # Record the number of itemsets and the time taken
    num_itemsets = len(frequent_itemsets)
    execution_time = end_time - start_time
    
    # Save the results for this min_freq
    results.append({
        'min_freq': min_freq,
        'num_itemsets': num_itemsets,
        'execution_time': execution_time
    })
    
    # Print summary for this min_freq
    print(f"min_freq: {min_freq} | Number of Itemsets: {num_itemsets} | Execution Time: {execution_time:.2f} seconds")
    print("-" * 60)

# Display all results
results_df = pd.DataFrame(results)
print(results_df)


Testing min_freq = 0.01


C:\Users\Owner\anaconda3\Lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


min_freq: 0.01 | Number of Itemsets: 31 | Execution Time: 0.52 seconds
------------------------------------------------------------
Testing min_freq = 0.005


C:\Users\Owner\anaconda3\Lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


min_freq: 0.005 | Number of Itemsets: 67 | Execution Time: 1.75 seconds
------------------------------------------------------------
Testing min_freq = 0.001


C:\Users\Owner\anaconda3\Lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


min_freq: 0.001 | Number of Itemsets: 211 | Execution Time: 14.82 seconds
------------------------------------------------------------
Testing min_freq = 0.0005


C:\Users\Owner\anaconda3\Lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


min_freq: 0.0005 | Number of Itemsets: 315 | Execution Time: 49.80 seconds
------------------------------------------------------------
Testing min_freq = 0.0001


C:\Users\Owner\anaconda3\Lib\site-packages\mlxtend\frequent_patterns\fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(
C:\Users\Owner\anaconda3\Lib\site-packages\executing\executing.py:713: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
  right=ast.Str(s=sentinel),
C:\Users\Owner\anaconda3\Lib\site-packages\executing\executing.py:713: DeprecationWarning: ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
  right=ast.Str(s=sentinel),
C:\Users\Owner\anaconda3\Lib\ast.py:587: DeprecationWarning: Attribute s is deprecated and will be removed in Python 3.14; use value instead
  return Constant(*args, **kwargs)
C:\Users\Owner\anaconda3\Lib\site-packages\executing\executing.py:713: DeprecationWarning: ast.Str is deprecated and will be remove

MemoryError: Unable to allocate 19.6 GiB for an array with shape (1230, 4, 533799) and data type int64

#### Analysis and Interpretation

A higher min-freq results in fewwer but more common itemsets. This seems useful for looking for frequently associated combinations of items.
A lower min-freq finds less frequent patterns, but does take more time to compute. I think that a good threshold would be between that two, which would give me a manageable and interesting number of itemsets while also keeping computation time down to a manageable level.